sdk/src/process_instruction.rs

```diff
@@ -15,19 +15,59 @@ use std::{cell::RefCell, fmt::Debug, rc::Rc, sync::Arc};
 /// invoke_context: Invocation context
 pub type LoaderEntrypoint = unsafe extern "C" fn(
     program_id: &Pubkey,
-    keyed_accounts: &[KeyedAccount],
     instruction_data: &[u8],
     invoke_context: &dyn InvokeContext,
 ) -> Result<(), InstructionError>;
 
 pub type ProcessInstructionWithContext =
-    fn(&Pubkey, &[KeyedAccount], &[u8], &mut dyn InvokeContext) -> Result<(), InstructionError>;
+    fn(&Pubkey, &[u8], &mut dyn InvokeContext) -> Result<(), InstructionError>;
+
+pub struct InvokeContextStackFrame<'a> {
+    pub key: Pubkey,
+    pub keyed_accounts: Vec<KeyedAccount<'a>>,
+    pub keyed_accounts_range: std::ops::Range<usize>,
+}
+
+impl<'a> InvokeContextStackFrame<'a> {
+    pub fn new(
+        key: Pubkey,
+        keyed_accounts: &[(&'a Pubkey, bool, bool, &'a RefCell<AccountSharedData>)],
+    ) -> Self {
+        let keyed_accounts: Vec<_> = keyed_accounts
```

Comment:
Not cloning here because we want to create new lifetimes?

Reply:
Never really liked that we have to create these with two different new functions.  I can see for `Accountmeta` making the distinction for developers but for `KeyedAccounts` it just looks messy.

I would recommend
- `.clone()` here unless I missing something
- replace new and new_readonly with just a new function that takes readonly as an argument

Reply:
I originally wanted to merge the two `KeyedAccount` constructors but was worried about breaking changes in the SDK / API. But, if you also think that it would be better that way, then I will do it so.

Reply:
We could only break backward compatibility at a major version change (https://docs.solana.com/developing/backwards-compatibility).  We could create a new function that takes writable as a parameter